Sources/Datadog/DatadogCore/Storage/FeatureStorage.swift

```diff
@@ -7,26 +7,101 @@
 import Foundation
 
 internal struct FeatureStorage {
-    /// Writes data to files. This `Writer` takes current value of the `TrackingConsent` into consideration
-    /// to decided if the data should be written to authorized or unauthorized folder.
-    let writer: AsyncWriter
-    /// Reads data from files in authorized folder.
-    let reader: SyncReader
+    /// The name of this Feature, used to distinguish storage instances in telemetry and logs.
+    let featureName: String
+    /// Queue for performing all I/O operations (writes, reads and files management).
+    let queue: DispatchQueue
+    /// Directories for managing data in this Feature.
+    let directories: FeatureDirectories
+    /// Orchestrates files collected in `.granted` consent.
+    let authorizedFilesOrchestrator: FilesOrchestratorType
+    /// Orchestrates files collected in `.pending` consent.
+    let unauthorizedFilesOrchestrator: FilesOrchestratorType
+    /// Encryption algorithm applied to persisted data.
+    let encryption: DataEncryption?
 
-    /// An arbitrary `Writer` which always writes data to authorized folder.
-    /// Should be only used by components which implement their own consideration of the `TrackingConsent` value
-    /// associated with data written (e.g. crash reporting integration which saves the consent value along with the crash report).
-    let arbitraryAuthorizedWriter: AsyncWriter
+    func writer(for trackingConsent: TrackingConsent) -> Writer {
+        switch trackingConsent {
+        case .granted:
+            return AsyncWriter(
+                execute: FileWriter(orchestrator: authorizedFilesOrchestrator, encryption: encryption),
+                on: queue
+            )
+        case .notGranted:
+            return NOPWriter()
+        case .pending:
+            return AsyncWriter(
+                execute: FileWriter(orchestrator: unauthorizedFilesOrchestrator, encryption: encryption),
+                on: queue
+            )
+        }
+    }
+
+    var reader: Reader {
+        DataReader(
+            readWriteQueue: queue,
+            fileReader: FileReader(
+                orchestrator: authorizedFilesOrchestrator,
+                encryption: encryption
+            )
+        )
+    }
+
+    func migrateData(fromConsent: TrackingConsent, toConsent: TrackingConsent) {
+        queue.async {
+            do {
+                switch (fromConsent, toConsent) {
+                case (.pending, .notGranted):
+                    try directories.unauthorized.deleteAllFiles()
+                case (.pending, .granted):
+                    try directories.unauthorized.moveAllFiles(to: directories.authorized)
+                default:
+                    break
+                }
```

Comment:
With the implementation of `func writer(for trackingConsent: TrackingConsent) -> Writer`, we don't need to know the previous consent. We can only rely on the new one to perform file operations.

Reply:
👍 done ✅